### ISSUE TITLE
New `geoip` transform

### ISSUE THREAD

**Issue by mre** [timberio/vector#847]
In our logstash pipelines we have quite a few calls to a GeoIP database.
See the [logstash filter for more info](https://www.elastic.co/guide/en/logstash/current/plugins-filters-geoip.html).
It would be great to have the same functionality in vector.

## Alternatives

* This might also be possible with Lua, although I haven't tried it. From what I can tell, the geoip modules for Lua are not seeing a lot of activity. E.g. [lua-geoip](https://github.com/agladysh/lua-geoip) was last updated in May 2017. 
* Another alternative might be to support a generic WASM runtime as proposed in https://github.com/timberio/vector/issues/564#issuecomment-507884581, although I'm not sure if we should go down that road for a very general transform like this. It feels like a lot of people might need this.

**Comment by mre** [comment]
An alternative might be [maxminddb-rust](https://github.com/oschwald/maxminddb-rust). Seems to be actively maintained and used by a few GeoIP webservices written in Rust ([here](https://github.com/mroth/nanogeoip/blob/master/Cargo.toml#L16) and [here](https://github.com/ffissore/geoip-rs/blob/master/Cargo.toml#L18)).

[Here](https://github.com/oschwald/maxminddb-rust/blob/master/examples/lookup.rs) is a usage example:

```rust
extern crate maxminddb;

use std::net::IpAddr;
use std::str::FromStr;

use maxminddb::geoip2;

fn main() {
    let reader =
        maxminddb::Reader::open_readfile("/usr/local/share/GeoIP/GeoIP2-City.mmdb").unwrap();
    let ip: IpAddr = FromStr::from_str("128.101.101.101").unwrap();
    let city: geoip2::City = reader.lookup(ip).unwrap();
    println!("{:?}", city);
}
```

**Comment by lukesteensen** [comment]
Definitely agree that this would be a great feature to have and makes sense to do as a native transform.

The maxminddb crate looks great. I'd be curious to benchmark performance and memory usage, but my main question would be about the database files themselves. It seems like the options would be:

1. Distribute them with vector
2. Introduce functionality in vector to fetch them
3. Document that the user should fetch them independently of vector

We could certainly start with (3) to get a working transform. I'm not sure if that would be a satisfying user experience or not. It seems like there are [dedicated tools](https://github.com/maxmind/geoipupdate) for fetching and update the database files, so maybe we leave it there?

**Comment by MarkusH** [comment]
License wise, I think the only sensible way is option 3, to document that a user needs to get the database files from somewhere. And with that, essentially making the path to the database file configurable.

**Comment by mre** [comment]
@lukesteensen [here](https://github.com/mre/maxminddb-rust-bench) is a quick benchmark using criterion. It should be fast enough. The gist is that it can look up 30.000 IPs per second on my machine (MacBook Pro 2015, i7, 16GB RAM). If that's not enough we can use rayon for an almost linear performance improvement. Then it's doing 100.000 lookups/s over here.

I agree with both of you that we should start with option 3.

**Comment by lukesteensen** [comment]
@mre Nice! That seems plenty fast for me. Do you know if it's reading the whole database into memory? I know some implementations memory map the file. I'm not terribly concerned about memory usage as long as it only affects users who have configured this specific transform, but it's still good to know.

**Comment by amitsaha** [comment]
hey folks, Are one of you working on this feature? I am keen to work on it, if none is.

**Comment by binarylogic** [comment]
@amitsaha we don't have anyone on the team currently working on this, but I can't speak for @MarkusH or @mre. We'd very much welcome a pull request.

**Comment by mre** [comment]
@amitsaha from my side go ahead. 😃

**Comment by mre** [comment]
@lukesteensen since version 0.11, maxminddb-rust offers an optional `mmap` feature flag for memory mapping. See the changelog [here](https://github.com/oschwald/maxminddb-rust/blob/master/CHANGELOG.md#0110---2018-11-12).

**Comment by MarkusH** [comment]
> hey folks, Are one of you working on this feature? I am keen to work on it, if none is.

@amitsaha please ahead. I'm not working on it.

**Comment by amitsaha** [comment]
Thanks folks. I will give it a shot.

**Comment by mre** [comment]
That sounds very reasonable to me. I would probably do exactly the same.
As for the input field, maybe we could name it `source` to stay consistent with how Logstash calls it. Would make transitioning a little easier.
The database option is called `database` over there. 
The target option is called `target` and the default value is `geoip`.

[Here](https://www.elastic.co/guide/en/logstash/current/plugins-filters-geoip.html#plugins-filters-geoip-options) is the reference.
I think the remaining options could be omitted for now.

**Comment by binarylogic** [comment]
@amitsaha let us know if you have any questions. You can also post your questions in https://spectrum.chat/vector.

**Comment by amitsaha** [comment]
hi @binarylogic thanks - will keep that in mind. I have started working on the patch, but hasn't spent a lot of time on it yet to make a PR.

**Comment by amitsaha** [comment]
hey folks, just wanted to get feedback regarding writing tests for this. Do we test against a "real" maxmind DB file? If so, where does it live (as far as CI/local testing goes)?

**Comment by lukesteensen** [comment]
@amitsaha Good question. If there is a small file available with an appropriate license, we could include it in a `test_data` directory or something like that. That does seem like it would be useful for testing.

**Comment by moschroe** [comment]
I have implemented an experimental, rough version of geoip support for a private project. Sadly, I never managed to find the time to open this issue and start a PR :/

I looked at logstash's feature set and felt that being able to configure every possible way of quoting every possible field of a lookup into every possible event field smelt rather turing complete.

I opted to integrate this with lua so all that flexibility would already be there. I created an rlua function that would do the lookup and a conversion in rust for `maxminddb::geoip2::City`, using `rlua_serde::to_value` to make it accessible from lua scripts.

Also ported the experimental feature set of `Ipv6Addr` for detecting non-public IPs and exposed this as an rlua function. That allows a lua script to perform the check and insert a "PRIV" country code or similar, enabling downstream filtering and distinction from cases where a lookup fails to yield a result.
